zkevm-circuits/src/evm_circuit/util/math_gadget/binary_number.rs

```diff
@@ -0,0 +1,70 @@
+use eth_types::Field;
+use gadgets::{binary_number::AsBits, util::Expr};
+use halo2_proofs::{
+    circuit::Value,
+    plonk::{Error, Expression},
+};
+
+use crate::evm_circuit::util::{constraint_builder::ConstraintBuilder, CachedRegion, Cell};
+
+#[derive(Clone, Debug)]
+pub struct BinaryNumberGadget<F, const N: usize> {
+    bits: [Cell<F>; N],
+}
+
+impl<F: Field, const N: usize> BinaryNumberGadget<F, N> {
+    pub(crate) fn construct(cb: &mut ConstraintBuilder<F>, value: Expression<F>) -> Self {
+        let bits = array_init::array_init(|_| cb.query_bool());
+
+        // each assigned bit should be boolean.
+        for bit in bits.iter() {
+            cb.require_boolean("each bit is 0 or 1", bit.expr());
+        }
```

Comment:
`query_bool` already constraints the bit to be boolean. So this implies double-constraining the value.

See: 
```rust
pub(crate) fn query_bool(&mut self) -> Cell<F> {
        let cell = self.query_cell();
        self.require_boolean("Constrain cell to be a bool", cell.expr());
        cell
    }
```

Reply:
Yes, I must have missed this :)